File: core-utils/src/test/java/io/appmetrica/analytics/coreutils/internal/cache/DataCacheTest.java
package io.appmetrica.analytics.coreutils.internal.cache;

import androidx.annotation.NonNull;
import io.appmetrica.analytics.testutils.CommonTest;
import java.util.concurrent.TimeUnit;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class DataCacheTest extends CommonTest {

    class TestDataCache extends DataCache<Object> {

        public TestDataCache(long refreshTime, long expiryTime, @NonNull String description) {
            super(refreshTime, expiryTime, description);
        }

        @Override
        protected boolean shouldUpdate(@NonNull Object newData) {
            return true;
        }
    }

    private final String description = "Some cache description";

    @Mock
    LocationDataCacheUpdateScheduler mDataCacheUpdateScheduler;

    private DataCache<Object> mDataCache;
    @Mock
    private Object mData;

    private final long defaultCacheRefreshTime = TimeUnit.MINUTES.toMillis(1);
    private final long defaultCacheExpiryTime = TimeUnit.MINUTES.toMillis(2);

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        mDataCache = new TestDataCache(defaultCacheRefreshTime, defaultCacheExpiryTime, description);
    }

    @Test
    public void testUpdateData() {
        mDataCache.updateData(mData);
        assertThat(mDataCache.getData()).isEqualTo(mData);
    }

    @Test
    public void testUpdateDataNullScheduler() {
        mDataCache.updateData(mData);
    }

    @Test
    public void testUpdateDataTriggersScheduler() {
        mDataCache.setUpdateScheduler(mDataCacheUpdateScheduler);
        mDataCache.updateData(mData);
        verify(mDataCacheUpdateScheduler).onStateUpdated();
    }

    @Test
    public void testGetData() {
        mDataCache.setUpdateScheduler(mDataCacheUpdateScheduler);
        mDataCache.updateData(mData);
        assertThat(mDataCache.getData()).isEqualTo(mData);
        verify(mDataCacheUpdateScheduler, never()).scheduleUpdateIfNeededNow();
    }

    @Test
    public void testGetDataIfShouldUpdate() throws InterruptedException {
        long refreshTime = 1000L;
        long invalidateDelta = 1000L;
        long accuracy = 100L;
        mDataCache = new TestDataCache(refreshTime, refreshTime + invalidateDelta, description);
        mDataCache.setUpdateScheduler(mDataCacheUpdateScheduler);
        mDataCache.updateData(mData);
        Thread.sleep(refreshTime + accuracy);
        assertThat(mDataCache.getData()).isEqualTo(mData);
        verify(mDataCacheUpdateScheduler).scheduleUpdateIfNeededNow();
    }

    @Test
    public void testGetDataIfShouldClear() throws InterruptedException {
        long refreshTime = 10L;
        long expiryTime = 12L;
        mDataCache = new TestDataCache(refreshTime, expiryTime, description);
        mDataCache.setUpdateScheduler(mDataCacheUpdateScheduler);
        mDataCache.updateData(mData);
        Thread.sleep(expiryTime + 1);
        assertThat(mDataCache.getData()).isNull();
        verify(mDataCacheUpdateScheduler).scheduleUpdateIfNeededNow();
    }

    @Test
    public void testGetDataIfCacheExpiredAndMissedUpdater() throws InterruptedException {
        mDataCache = new TestDataCache(10, 20, description);
        mDataCache.updateData(mData);
        Thread.sleep(10);
        mDataCache.getData();
    }

    @Test
    public void testGetDataIfCacheIsEmpty() {
        mDataCache.setUpdateScheduler(mDataCacheUpdateScheduler);
        mDataCache.getData();
        verify(mDataCacheUpdateScheduler).scheduleUpdateIfNeededNow();
    }

    @Test
    public void testShouldUpdateIfCacheIsEmpty() {
        assertThat(mDataCache.shouldUpdate()).isTrue();
    }

    @Test
    public void testShouldUpdateCacheIsOutdated() {
        long refreshTime = 10L;
        mDataCache = new TestDataCache(refreshTime, 20, description);
        mDataCache.updateData(mData);
        try {
            Thread.sleep(refreshTime + 1);
        } catch (InterruptedException e) {}
        assertThat(mDataCache.shouldUpdate()).isTrue();
    }

    @Test
    public void testShouldNotUpdate() {
        mDataCache.updateData(mData);
        assertThat(mDataCache.shouldUpdate()).isFalse();
    }

    @Test
    public void updateCacheControl() {
        long refreshTime = 123545;
        long expiryTime = 2312321L;
        mDataCache.updateCacheControl(refreshTime, expiryTime);
        assertThat(mDataCache.getCachedData().getRefreshTime()).isEqualTo(refreshTime);
        assertThat(mDataCache.getCachedData().getExpiryTime()).isEqualTo(expiryTime);
    }
}

File: core-utils/src/test/java/io/appmetrica/analytics/coreutils/internal/cache/DataCacheTest.kt
package io.appmetrica.analytics.coreutils.internal.cache

import io.appmetrica.analytics.coreutils.internal.time.SystemTimeProvider
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.MockedConstructionRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner
import java.util.concurrent.TimeUnit

@RunWith(RobolectricTestRunner::class)
class DataCacheTest : CommonTest() {

    inner class TestDataCache(
        refreshTime: Long,
        expiryTime: Long,
        description: String
    ) : DataCache<Any>(refreshTime, expiryTime, description) {
        override fun shouldUpdate(newData: Any): Boolean = true
    }

    private val description = "Some cache description"

    private val dataCacheUpdateScheduler: LocationDataCacheUpdateScheduler = mock()

    private lateinit var dataCache: DataCache<Any>

    private val data: Any = mock()

    private val defaultCacheRefreshTime = TimeUnit.MINUTES.toMillis(1)
    private val defaultCacheExpiryTime = TimeUnit.MINUTES.toMillis(2)

    @get:Rule
    val timeProviderRule = MockedConstructionRule(SystemTimeProvider::class.java)

    @get:Rule
    val cachedDataRule = MockedConstructionRule(CachedDataProvider.CachedData::class.java)

    @Before
    fun setUp() {
        dataCache = TestDataCache(defaultCacheRefreshTime, defaultCacheExpiryTime, description)
    }

    private fun cachedData(): CachedDataProvider.CachedData<Any> {
        assertThat(cachedDataRule.constructionMock.constructed()).isNotEmpty
        val constructed = cachedDataRule.constructionMock.constructed()
        @Suppress("UNCHECKED_CAST")
        return constructed[constructed.size - 1] as CachedDataProvider.CachedData<Any>
    }

    @Test
    fun updateData() {
        dataCache.updateData(data)

        verify(cachedData()).setData(data)
    }

    @Test
    fun updateDataNullScheduler() {
        dataCache.updateData(data)

        verify(cachedData()).setData(data)
    }

    @Test
    fun updateDataTriggersScheduler() {
        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.updateData(data)

        verify(cachedData()).setData(data)
        verify(dataCacheUpdateScheduler).onStateUpdated()
    }

    @Test
    fun getData() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(data)

        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.updateData(data)

        assertThat(dataCache.data).isEqualTo(data)
        verify(dataCacheUpdateScheduler, never()).scheduleUpdateIfNeededNow()
    }

    @Test
    fun getDataIfShouldUpdate() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(true)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(data)

        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.updateData(data)

        assertThat(dataCache.data).isEqualTo(data)
        verify(dataCacheUpdateScheduler).scheduleUpdateIfNeededNow()
    }

    @Test
    fun getDataIfShouldClear() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)
        whenever(cachedData().shouldClearData()).thenReturn(true)
        whenever(cachedData().data).thenReturn(null)

        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.updateData(data)

        assertThat(dataCache.data).isNull()
        verify(cachedData()).setData(null)
    }

    @Test
    fun getDataIfCacheExpiredAndMissedUpdater() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(true)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(data)

        dataCache.updateData(data)
        dataCache.data

        // No scheduler, so scheduleUpdateIfNeededNow should not be called
    }

    @Test
    fun getDataIfCacheIsEmpty() {
        whenever(cachedData().isEmpty).thenReturn(true)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(null)

        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.data

        verify(dataCacheUpdateScheduler).scheduleUpdateIfNeededNow()
    }

    @Test
    fun shouldUpdateIfCacheIsEmpty() {
        whenever(cachedData().isEmpty).thenReturn(true)

        assertThat(dataCache.shouldUpdate()).isTrue
    }

    @Test
    fun shouldUpdateCacheIsOutdated() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(true)

        dataCache.updateData(data)

        assertThat(dataCache.shouldUpdate()).isTrue
    }

    @Test
    fun shouldNotUpdate() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)

        dataCache.updateData(data)

        assertThat(dataCache.shouldUpdate()).isFalse
    }

    @Test
    fun updateCacheControl() {
        val refreshTime = 123545L
        val expiryTime = 2312321L

        dataCache.updateCacheControl(refreshTime, expiryTime)

        verify(cachedData()).setExpirationPolicy(refreshTime, expiryTime)
    }

    @Test
    fun shouldUpdateWhenTimeGoesBackwards() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(true)

        dataCache.updateData(data)

        assertThat(dataCache.shouldUpdate()).isTrue
    }

    @Test
    fun shouldClearWhenTimeGoesBackwards() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)
        whenever(cachedData().shouldClearData()).thenReturn(true)
        whenever(cachedData().data).thenReturn(null)

        dataCache.updateData(data)

        assertThat(dataCache.data).isNull()
        verify(cachedData()).setData(null)
    }

    @Test
    fun dataWithinRefreshTime() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(data)

        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.updateData(data)

        assertThat(dataCache.data).isEqualTo(data)
        verify(dataCacheUpdateScheduler, never()).scheduleUpdateIfNeededNow()
    }

    @Test
    fun dataExactlyAtRefreshTime() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(data)

        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.updateData(data)

        assertThat(dataCache.data).isEqualTo(data)
        // At exactly refreshTime, should NOT update (only > refreshTime)
        verify(dataCacheUpdateScheduler, never()).scheduleUpdateIfNeededNow()
    }

    @Test
    fun dataJustPastRefreshTime() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(true)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(data)

        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.updateData(data)

        assertThat(dataCache.data).isEqualTo(data)
        verify(dataCacheUpdateScheduler).scheduleUpdateIfNeededNow()
    }

    @Test
    fun dataExactlyAtExpiryTime() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(data)

        dataCache.updateData(data)

        assertThat(dataCache.data).isEqualTo(data)
        // At exactly expiryTime, should NOT clear (only > expiryTime)
    }

    @Test
    fun dataJustPastExpiryTime() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)
        whenever(cachedData().shouldClearData()).thenReturn(true)
        whenever(cachedData().data).thenReturn(null)

        dataCache.updateData(data)

        assertThat(dataCache.data).isNull()
        verify(cachedData()).setData(null)
    }
}
